src/desktop/Window.cpp

```diff
@@ -856,21 +856,53 @@ bool CWindow::isTagged(const std::string& tag, bool strict) {
 }
 
 void CWindow::applyTag(const std::string& tag, bool dynamic) {
-    bool        isSet    = true;
+
     std::string tag_real = dynamic ? tag + "*" : tag;
 
-    if (tag.starts_with("-")) { // unset
-        isSet    = false;
+    bool        tagged  = isTagged(tag_real, true);
+    bool        changed = true;
+    bool        setTag  = true;
+
+    if (tag_real.starts_with("-")) { // unset
         tag_real = tag_real.substr(1);
-    } else if (tag.starts_with("+")) // set
+        changed  = tagged;
+        setTag   = false;
+    } else if (tag_real.starts_with("+")) { // set
         tag_real = tag_real.substr(1);
-    else // toggle if without prefix
-        isSet = !isTagged(tag_real, true);
+        changed  = !tagged;
+    } else // toggle if without prefix
+        setTag = !tagged;
 
-    if (isSet)
+    if (!changed)
+        return;
+
+    if (setTag)
         m_tags.emplace(tag_real);
     else
         m_tags.erase(tag_real);
+
+    // re run tag rules after update tag
+    queueScanForTagMatch();
+}
+
+void CWindow::queueScanForTagMatch() {
+    // TODO: assume single thread for window operation
+    if (!m_tagScanSource)
+        m_tagScanSource = wl_event_loop_add_idle(
+            g_pCompositor->m_sWLEventLoop,
+            [](void* data) {
+                auto w = static_cast<CWindow*>(data);
+
+                w->m_vMatchedRules = g_pConfigManager->getMatchingRules(w->m_pSelf.lock());
+                for (auto& r : w->m_vMatchedRules)
+                    if ((r.v2 && !r.szTag.empty() && w->isTagged(r.szTag)) || (!r.v2 && r.szValue.starts_with("tag:") && w->isTagged(r.szValue.substr(4))))
+                        w->applyDynamicRule(r);
+
+                // NOTE: if tag from tag, then the nested tag's rescan will be ignored
+                wl_event_source_remove(w->m_tagScanSource);
+                w->m_tagScanSource = nullptr;
```

Comment:
reset `m_tagScanSource` at the end of callback to avoid tag toggle from tag match, which will cause infinite loop in such case.

Reply:
For a case like this, it's reasonable:
```
windowrulev2 = tag +code, tag:cc
windowrulev2 = tag +code, tag:go
windowrulev2 = tag +code, tag:py

windowrulev2 = bordercolor rgba(800000cc), tag:cc
windowrulev2 = bordercolor rgba(008000cc), tag:go
windowrulev2 = bordercolor rgba(000080cc), tag:py

windowrulev2 = opacity 0.7, tag:code
```

While, with latest commit, the matched rules returned from config is already a full match with tag rule considered, there is no need to queue rescan in `applyTag()`

Reply:
It might be reasonable for a case like this:

```
windowrulev2 = tag +code, tag:cpp
windowrulev2 = tag +code, tag:go
windowrulev2 = tag +code, tag:py

windowrulev2 = bordercolor rgba(800000cc), tag:cpp
windowrulev2 = bordercolor rgba(008000cc), tag:go
windowrulev2 = bordercolor rgba(000080cc), tag:py

windowrulev2 = opacity 0.8 override, tag:code
windowrulev2 = opacity 0.4 override, tag:cpp
```

Reply:
why not just assign 2 tags?

Reply:
Dynamic tags is more to be automatic, static tag is generally assigned manually. With dynamic tag, static tags can be grouped, or let me say, tags can be tagged too. It's like abstract class and sub subclass to some degree, and currently this does not bring more footprint than a local tag emulation in matching tags. the general use case is documented in https://github.com/hyprwm/hyprland-wiki/pull/660

Reply:
mhm ok